Initial design ideas for Source

pivotal-nader-ziada opened the pull request:
Signed-off-by: Shash Reddy <shashwathireddy@gmail.com>

Design proposal for #11 #12 and #13 

## Proposed Changes

  * Create a PipelineSource and SourceVersion
  * remove ArtifactSourceBinding and use SourceBinding as both Input and Output
  * Create an interface for Sources
  * `git` and `image` implementation of Source
  * `passedConstraint` to make sure a task uses the Source Version that passed a previous task

This PR is just to gather feedback about these design ideas.

ImJasonH commented on pkg/apis/pipeline/v1beta1/git_source.go:
```diff
@@ -0,0 +1,45 @@
+	Name           string `json:"name"`
+	Type           string `json:"type"`
+	URL            string `json:"url"`
+	Branch         string `json:"branch"`
```
`Revision` ?

In Build's `GitSource` today, we decided to accept branches, tags and arbitrary refs (including commits) in the same field, using Git's built-in semantics.

pivotal-nader-ziada replied:
make sense to be consistent

tejal29 replied:
Few tips:
We probably define interface `Source`  method `Type`. 
type SourceType string
enum 
```
type Source interface {
    Type() SourceType
    Version() string 
   // Pulls source code.
    Pull() error 
}
```
And Then the `GitSource` can define its own fields like you have and implement methods
```
func (gs GitSource) Version() string {
    return gs.Commit
}

func (gs GitSource) Type() SourceType {
  return constants.GitSource
} 
```

tejal29 replied:
That ways, other sources can be added easily

ImJasonH commented on pkg/apis/pipeline/v1beta1/git_source.go:
```diff
@@ -0,0 +1,45 @@
+
+func (s GitSource) getParams() []Param {
+	var result []Param
+	return result
```
`return []Param{}`?

These could also be written on single lines to conserve vertical space:

```go
func (GitSource) getParams() []Param { return []Param{} }
```

pivotal-nader-ziada replied:
It really should also be returning the actual values of any existing Params

bobcatfish commented on pkg/apis/pipeline/v1beta1/sourceversion_types.go:
```diff
@@ -0,0 +1,57 @@
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SourceVersion is the Schema for the sourceversion API
```
maybe a bit here about how a `SourceVersion` would be intended to be used?

pivotal-nader-ziada replied:
SourceVersion is for the `taskrun_controller` to make sure the same version of the Source (eg: git commit) is used across the same pipeline run, will update the comment

bobcatfish commented on pkg/apis/pipeline/v1beta1/git_source.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func (s GitSource) getVersion() string {
+	return s.Commit
```
ah cool, so _anything_ could have a version potentially, and for a git source, it happens to be the commit - pretty cool!

pivotal-nader-ziada replied:
yeah, exactly, it will mean different things for different types

tejal29 replied:
+ 1. Just a note, lets lock down a Source interface and then GitSource implement the interface.

bobcatfish commented on pkg/apis/pipeline/v1beta1/image_source.go:
```diff
@@ -0,0 +1,43 @@
+	Name string `json:"name"`
+	// TODO: maybe an enum, with values like 'registry', GCS bucket
+	Type string `json:"type"`
+	URL  string `json:"url"`
```
I think it would maybe be useful to separate the endpoint where the image is tored from the URL - we might want to have an endpoint, tag, and sha all available somehow

shashwathi replied:
Thats would be the metadata for image source. We could consider `metadata` for source's interface to provide this information.

tejal29 replied:
How about we use the subset of fields Redhat Atomic Host Signature. https://github.com/aweiteka/image/blob/e5a20d98fe698732df2b142846d007b45873627f/docs/signature.md 
```
 "identity": {
            "docker-reference": "busybox"
        },
        "image": {
            "docker-manifest-digest": "sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6"
        }, 
```
Some Missing fields are "Registry"

shashwathi replied:
`docker-reference` could be combination of `registry:docker-image-name` as well.

bobcatfish commented on examples/pipelines/kritis.yaml:
```diff
@@ -47,9 +45,24 @@ spec:
           params:
               - name: testArgs
                 value: "-e REMOTE_INTEGRATION=true"
+    sources:
```
in the current design, this would make sense to put into the [PipelineParams](https://github.com/knative/build-pipeline#pipelineparams), i.e. replacing [these sources in the example](https://github.com/knative/build-pipeline/blob/master/examples/pipelineparams.yaml#L8)

pivotal-nader-ziada replied:
I wanted to make them outside of the `tasks` so it can be used across multiple tasks, but it would work either outside or in the PipelineParams

bobcatfish replied:
I think we'd definitely want this in PipelineParams, unless we decide in the long run to do away with PipelineParams

bobcatfish commented on pkg/apis/pipeline/v1beta1/pipeline_types.go:
```diff
@@ -101,6 +95,24 @@ type TaskRef struct {
+}
+
+// SourceRef can be used to refer to a specific instance of a Source.
+type SourceRef struct {
```
So is `SourceRef` the interface that `Sources` must comply with? Would a `SourceRef` be expected to have the functions that the git + image types above have? e.g. `getVersion`?

pivotal-nader-ziada replied:
i think so :)

bobcatfish commented:
> remove ArtifactSourceBinding and use SourceBinding as both Input and Output

This is a cool idea, and more like some of the coolness that concourse has - it makes a lot of sense, since the line between the two is so fuzzy anyway e.g. images could be either inputs or outputs.

+ @tejal29 who is driving the artifacts and result store portion of the design, this would change the scope a bit.

> passedConstraint to make sure a task uses the Source Version that passed a previous task

Interesting!! So this would actually tackle #11 as well.

It seems like this changes the model a bit: instead of passing a `Source`/image from task to task, you can declare it at the top level and then declare what stages it needs to pass :thinking: 

I feel like this makes the model a bit more complicated: we now have one more way of expressing the DAG (nextTask, prevTask, paramBindings AND passedConstraints), unless there's a way we can do away with paramBindings completely and replace them with passedConstraints? I'm not completely seeing the advantage of using passedConstraints over using paramBindings.

Would be helpful to see a list of pros and cons of both!

> git and image implementation of Source

Do you have any thoughts about what one would do if they wanted to extend this model? I guess it would require rebuilding the controller, but you could extend it with anything that complies with the interface?

As we move forward it would also be great to see what this looks like in a [TaskRun](https://github.com/knative/build-pipeline#taskrun) (e.g. [the kritis task run](https://github.com/knative/build-pipeline/blob/master/examples/invocations/run-kritis-test.yaml))

My main questions:
- Would like to be convinced that sourcebindings + passedConstraints are > paramBindings
- Would I have to declare a Source for any image I intend to build and pass between Tasks in my Pipeline? (maybe that's a good thing!)

pivotal-nader-ziada commented:
> This is a cool idea, and more like some of the coolness that concourse has - it makes a lot of sense, > since the line between the two is so fuzzy anyway e.g. images could be either inputs or outputs.

yes images can be either inputs and outputs, maybe the name `Source` is not the best with it representing inputs and outputs. This would it make it easier to pass an image from the build to the test task.

> Interesting!! So this would actually tackle #11 as well.

Like you said this would address #11 as well to make it easier for the code to track which Source is moving along between tasks in a PipelineRun

> It seems like this changes the model a bit: instead of passing a Source/image from task to task, you > can declare it at the top level and then declare what stages it needs to pass 🤔

You would have to declare any `image` (Source) at the top level and tasks can use it as input or output, each image would have a different `Source` top level item. The `passed` is to enforce the image is not latest from that source repo (or registry) but the one that got built

> I feel like this makes the model a bit more complicated: we now have one more way of expressing the > DAG (nextTask, prevTask, paramBindings AND passedConstraints), unless there's a way we can do 
> away with paramBindings completely and replace them with passedConstraints? I'm not completely > seeing the advantage of using passedConstraints over using paramBindings.

you are correct, we might not need the ParamBinding, the reason I added the passedConstraint is that it makes the dependency of artifacts more explicit, didn't know how ParamBindings can do that?

> Do you have any thoughts about what one would do if they wanted to extend this model? I guess it 
> would require rebuilding the controller, but you could extend it with anything that complies with the 
> interface?

Not sure if we have to add something, but I though anything that implements the interface should work, I can try it out

> Would like to be convinced that sourcebindings + passedConstraints are > paramBindings

will work on some more examples of the pipeline yaml to demonstrate it more

> Would I have to declare a Source for any image I intend to build and pass between Tasks in my Pipeline? (maybe that's a good thing!) 

yes, each image needs a Source

This PR is just to start the conversation about the `Source` part of the pipeline, and I agree it increases the scope a bit, but I think it provides a lot of flexibility and value going forward, its similar to the `resource` concept in Concourse which I have used a lot and appreciate the benefits of.

We are open to any modifications we come up while working on this together

shashwathi commented:
> I feel like this makes the model a bit more complicated: we now have one more way of expressing the > DAG (nextTask, prevTask, paramBindings AND passedConstraints), unless there's a way we can do
away with paramBindings completely and replace them with passedConstraints? I'm not completely > seeing the advantage of using passedConstraints over using paramBindings.

Adding to @pivotal-nader-ziada's  point about the advantage of `passedConstraints`. I think it also will describes granular dependency across sources for more accurate DAG than using `prevTask`, `nextTask` which provide generic task dependency.

For example

```
apiVersion: pipeline.knative.dev/v1beta1
kind: Pipeline
metadata:
  name: passed-pipeline
  namespace: default
spec:
    tasks:
        - name: unit-test-kritis          # 1.  Run unit Tests
          taskRef:
              name: make
          inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis
         inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis-test-script
          params:
              - name: makeTarget
                value: test
        - name: integration-tests-kritis               # 2. Run integration Tests
          taskRef:
              name: build-push
          inputSourceBindings:
              - name: workspace
                sourceKey: kritis
          passedConstraint: [unit-test-kritis]
          inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis-test-script
          params:
              - name: makeTarget
                value: integration-test
    sources:
      - name: kritis
        type: git
        params:
          - name: url
            value: https://github.com/grafeas/kritis
          - name: branch
            value: master
    - name: kritis-test-script
        type: git
        params:
          - name: url
            value: https://github.com/grafeas/kritis-test-scripts
          - name: branch
            value: master
```

It provides information that task `integration-tests-kritis`.source(kritis) is dependent on previous task but `integration-tests-kritis ` is okay to pull latest commit on `kritis-test-script`. This is a common use case as a release author, I would like same version of application code passing through the pipeline but not the test code. This makes the dependency of sources across tasks very explicit. Another advantage is the fan-in and fan-out model is supported only with `passedConstraints`. I will add to the above example to demonstrate

```
apiVersion: pipeline.knative.dev/v1beta1
kind: Pipeline
metadata:
  name: passed-pipeline
  namespace: default
spec:
    tasks:
        - name: unit-test-kritis          # 1.  Run unit Tests
          taskRef:
              name: make
          inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis
         inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis-test-script
          params:
              - name: makeTarget
                value: test
        - name: integration-tests-kritis               # 2. Run integration Tests
          taskRef:
              name: build-push
          inputSourceBindings:
              - name: workspace
                sourceKey: kritis
          passedConstraint: [unit-test-kritis]            # fan-out after unit testing
          inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis-test-script
          params:
              - name: makeTarget
                value: integration-test
   - name: stress-tests-kritis               # 2. Run stress Tests
          taskRef:
              name: build-push
          inputSourceBindings:
              - name: workspace
                sourceKey: kritis
          passedConstraint: [unit-test-kritis]            # fan-out after unit testing
          inputSourceBindings:
              - inputName: workspace
                sourceKey: kritis-test-script
          params:
              - name: makeTarget
                value: stress-test
          taskRef:       
              name: build-push           # 3. Build and push
          inputSourceBindings:
              - name: workspace
                sourceKey: kritis
         passedConstraint: [stress-test-kritis, integration-test-kritis]  # fan-in after all testing
          outputSourceBindings:
              - name: registry
                sourceKey: stagingRegistry
          params:
              - name: pathToDockerfile
                value: deploy/Dockerfile
    sources:
    ....
```

WDYT? @pivotal-nader-ziada @bobcatfish @ImJasonH

tejal29 commented on pkg/apis/pipeline/v1beta1/pipelineparams_types.go:
```diff
@@ -22,21 +22,16 @@ import (
-	Branch         string `json:"branch"`
-	Commit         string `json:"commit,omitempty"`
-	ServiceAccount string `json:"serviceAccount,omitempty"`
+type Source interface {
```
Ah i see you have defined Source interface here.

tejal29 commented on pkg/apis/pipeline/v1beta1/pipelineparams_types.go:
```diff
@@ -22,21 +22,16 @@ import (
+	getType() string
+	getParams() []Param
+	getVersion() string
+	getPassingConstraints() []TaskRef
```
Can you add more information on this method?
getPassingConstraints()

pivotal-nader-ziada replied:
good catch, that should not be here.  We first thought about putting the constraint about a Source being passed from one task to another here, but moved it to SourceBinding. Missed removing it from here. Where do you think is a better place for that?

tejal29 replied:
I feel, `getPassingConstrainsts` should be defined at the interface level. 
The reason being, for each source type, they are going to be constant. You can multiple sources of same source type defined in your cluster. You do not want to repeat the binding information for each  sourcebinding.
Does that make sense?

bobcatfish commented:
Thanks for putting in those examples @shashwathi , those really help make the ideas more clear!

> This is a common use case as a release author, I would like same version of application code passing through the pipeline but not the test code.

I don't quite agree with this b/c I think the test code should be versioned along with the code it is testing and it would be very confusing to have a different commit's test code running.

The simplest would be to use the same version of the test code as the application code, and so my feeling is that by default a `source` should be used at the same version all the way through the pipeline.

> This makes the dependency of sources across tasks very explicit.

I agree! It's interesting that you have to declare all your sources up front as well, which I think provides more clarity, e.g. [in the previous kritis example](https://github.com/knative/build-pipeline/blob/master/examples/pipelines/kritis.yaml#L23) halfway through the pipeline there's suddenly this image being passed around, but if you just glance at the pipeline you have no idea it's producing image. I think your suggestion is a definite improvement here!

> Another advantage is the fan-in and fan-out model is supported only with passedConstraints.

Interesting! So it seems like we could potentially replace `nextTask`, `prevTask` AND `paramBindings` with `passedConstraints`!

Would it be possible to use `passedConstraints` even if a source is not involved?

pivotal-nader-ziada commented:
we will work on updating the PR with all the comments and adding a README with examples 
@shashwathi

bobcatfish commented:
Sounds great, thanks @pivotal-nader-ziada !!

> I don't quite agree with this b/c I think the test code should be versioned along with the code it is testing and it would be very confusing to have a different commit's test code running.

Quick update that @shashwathi brought to light some really important usecases here, e.g. in a super long running pipeline you need to make a change to a repo, e.g. a test, and don't want to have to re-run the whole thing.

shashwathi commented:
> Would it be possible to use passedConstraints even if a source is not involved?

What would be the use case for this?

bobcatfish commented:
> Would it be possible to use passedConstraints even if a source is not involved?
> What would be the use case for this?

It's hard to think of a good one, but here's the best that I've got: maybe you want to send a slack notification (via a Task) after a Task has run, e.g. after the integration tests pass, you send the message?

On the other hand, from today's discussion it sounds like we probably want to create some kind of notification specific design 🤔 so maybe this is a bad example.

bobcatfish commented:
@shashwathi @pivotal-nader-ziada after thinking a bit more about our discussion on Friday, I've created a couple follow up issues:

#49 - Designing notifications
#50 - Partial pipeline execution

After thinking more about the usecase where user wants to update a source needed by a long running Pipeline as the Pipeline is executing, I am thinking:
- Instead of pulling from HEAD, what if we implement #50 and allow the user to stop the Pipeline between Tasks and resume execution explicitly from a new commit?
- How often is a long running Pipeline broken up into many tasks? When I try to imagine use cases where a Pipeline is running for e.g. 15 hours, I can only think of cases where the long running portion is in one Task, e.g. a load test that runs over the weekend. In this case, being able to update HEAD of a repo 10 hours into the run wouldn't help, b/c the Task would have already started executing from a previous HEAD

pivotal-nader-ziada commented:
We (myself and @shashwathi) are in sync with the idea @ImJasonH explained in our discussion that a `PipelineRun` is the minimum unit we can run and that we are not expecting long running pipelines, or trying to avoid that at least. In that case, the `Partial pipeline execution` might be a lower priority for now.
We like the `notifications` idea

bobcatfish commented:
> PipelineRun is the minimum unit we can run and that we are not expecting long running pipelines, or trying to avoid that at least. In that case, the Partial pipeline execution might be a lower priority for now.

That's reasonable @pivotal-nader-ziada  - I'm fine with assuming that we'll avoid longer running pipelines for now, and revisiting this later.

What do you think @ImJasonH ? I'm fine with resolving #50 as won't fix for now, we could always bring it back later if we need it.

ImJasonH commented:
If nobody needs #50 now then I'm okay with thinking about it more later, but I still think we'll need it eventually. :+1:

tejal29 commented:
Sorry @shashwathi  jumping on this late. 
>Adding to @pivotal-nader-ziada's point about the advantage of passedConstraints. I think it also will describes granular dependency across sources for more accurate DAG than using prevTask, nextTask which provide generic task dependency.

Maybe we all (at least i am not :) ) are not on the same page what `passedConstraint` means. 
I was thinking `passedConstraint` as a way for sources to pass on to the next task "the version of code i just built". The Pipeline Controller can then pass this version to next task in the pipeline.
I do not see this as a way to provide task dependency. @pivotal-nader-ziada  what were you thinking initially?

I feel `passedConstraint` should be eliminated from config and it should be an Source interface method which the controller will use internally.

bobcatfish commented:
> Maybe we all (at least i am not :) ) are not on the same page what passedConstraint means.
I was thinking passedConstraint as a way for sources to pass on to the next task "the version of code i just built".

@pivotal-nader-ziada and @shashwathi can go into more detail but I understood `passedConstraints` to do two things:
1. Ensure that the "version" of a Source/Resource used for a Task is the same as the version used for the Tasks in `passedConstraints`
2. Express the DAG, i.e. if task B needs a Source/Resource that specifies `passedConstraints: [task A]` then task A would need to run before task B

Re. (1) my preference is that we should always use the same version of the Source/Resource regardless, but I think that (2) is cleaner than what we had before, and @shashwathi 's exmaple in https://github.com/knative/build-pipeline/pull/39#issuecomment-420669491 shows how it can be used for both fan in and fan out

> To make the config easier, we can also drop `inputSourceBindings` and `outputSourceBindings` and just have `sourceBindings`

Interesting idea!

At the moment, `inputSourceBindings` would specify data that needs to be mounted into containers run by the Task, and `outputSourceBindings` would declare outputs that the Task is expected to produce ([more discussion in Knative Build Artifacts Doc](https://docs.google.com/document/d/1FUn9LZXUAJQb8atoTeRxmsdjxpstOyegHF7ieJEblNs/edit)) (plus both would be available for templating). I'm not sure how we would make those work with just `sourceBindings`?

pivotal-nader-ziada commented:
@tejal29 @bobcatfish 

We are working on applying all the comments and adding examples that will make things more clear, coming soon (today)


> passedConstraints to do two things:

> Ensure that the "version" of a Source/Resource used for a Task is the same as the version used for > the Tasks in passedConstraints
> Express the DAG, i.e. if task B needs a Source/Resource that specifies passedConstraints: [task A] > then task A would need to run before task B
> Re. (1) my preference is that we should always use the same version of the Source/Resource > regardless, but I think that (2) is cleaner than what we had before, and @shashwathi 's exmaple in > #39 (comment) shows how it can be used for both fan in and fan out

yes, exactly, passed constraint would replace nextTask and prevTask and be used to drive the fan in and out as @bobcatfish explained above

I think the `intputSourceBinding` and `outputSourceBinding` just make the yaml more explicit and easier to read

tejal29 commented:
>yes, exactly, passed constraint would replace nextTask and prevTask and be used to drive the fan in and out as @bobcatfish explained above
In favor of replacing nextTask, prevTask by one keyword. I was playing with srewdriver.cd other day and i was going to propose `requires` as a construct instead. 
Link: https://docs.screwdriver.cd/user-guide/configuration/workflow#defining-workflow-order

> I think the intputSourceBinding and outputSourceBinding just make the yaml more explicit and easier to read.
Works!

shashwathi commented:
/check-cla

pivotal-nader-ziada commented:
@bobcatfish not sure why CLA is not working! seems the project maintainer can somehow fix that

bobcatfish commented:
> @bobcatfish not sure why CLA is not working! seems the project maintainer can somehow fix that

Sorry about that @pivotal-nader-ziada , we'll follow up and see what's going on!

bobcatfish commented on config/samples/pipeline_v1beta1_pipeline.yaml:
```diff
@@ -7,46 +7,60 @@ spec:
+    - name: stagingRegistry
+      resourceRef: 
+        name: stagingRegistry
+      type: imageRegistry
```
based on the changes we're looking at, should this be `type: image` instead?

pivotal-nader-ziada replied:
yes, type should be `image`, will fix that

tejal29 commented on examples/pipelines/guestbook.yaml:
```diff
@@ -8,76 +8,99 @@ spec:
           params:
               - name: pathToFiles
-                value: guestbook/all-in-one/guestbook-all-in-one.yaml
+                type: guestbook/all-in-one/guestbook-all-in-one.yaml
```
this is probably by mistake?

pivotal-nader-ziada replied:
this field was updated to type in PR #54, we used it to make the tests pass until we get a chance to ask about it

bobcatfish replied:
As far as I can tell, this should still be `value`, see [`PipelineTaskParam`](https://github.com/knative/build-pipeline/blob/801db1da0f0a483c29c9e9d30dea813281c05f0f/pkg/apis/pipeline/v1beta1/pipeline_types.go#L71).

pivotal-nader-ziada replied:
The tests were failing when it was ‘value’

bobcatfish commented on config/samples/pipeline_v1beta1_pipeline.yaml:
```diff
@@ -7,46 +7,60 @@ spec:
+      commit: HEAD
+    - name: stagingRegistry
+      resourceRef: 
+        name: stagingRegistry
```
i'm trying to understand the `resourceRef` in the context of the `resource` - should there  be a `stagingRef` CRD yaml that accompanies this to demonstrate what is in the CRD being referred to here?

I think it would be simpler to avoid having a separate CRD for resources specifically, but maybe there is some additional functionality that I'm missing!

pivotal-nader-ziada replied:
yeah, I had second thoughts about the resource ref also, I will consider removing it

tejal29 commented on examples/pipelines/guestbook.yaml:
```diff
@@ -8,76 +8,99 @@ spec:
+        name: stagingRegistry      
+      type: image
+      url: gcr.io/demo-staging
+      digest: latest
```
Having `digest` hard coded to `latest` will not work.  In case when you have multiple run of the same pipeline, the images will be over-written. 
We would want each pipeline run push a new image. The image digest should ideally correspond to the source version.

pivotal-nader-ziada replied:
yes, as mentioned above, will update the examples to clarify that this resource is the `image` and fix the fields for it. You are correct, we don't know the digest yet, that would be the resourceVersion in the taskRun

bobcatfish commented on config/samples/pipeline_v1beta1_pipeline.yaml:
```diff
@@ -7,46 +7,60 @@ spec:
+            - name: stagingRegistry
+              sourceKey: stagingRegistry
+              passedConstraint: [build-push-task]
+    resources:
```
what are your thoughts on declaring these resources in the Pipeline vs the PipelineParams?

I feel like they should be in the PipelineParams - or if not, that we should move everything that was in PipelineParams into Pipeline, but it seems like maybe we have different ideas about what makes sense in each?

pivotal-nader-ziada replied:
My preference for using the `Pipeline` is that the `pipelineParams` is like a generic bucket that had items not related to each other. I was going to propose moving all of the types currently in `pipelineParams` to `Pipeline` under their own heading. This way the yaml file is more clear to read and understand

pivotal-nader-ziada replied:
Yes i like the idea of separate yaml for resources. Allows you to isolate the different environments

tejal29 commented on examples/pipelines/kritis.yaml:
```diff
@@ -5,51 +5,67 @@ metadata:
+        revision: master
+      - name: kritis-test-github
+        resourceRef:
+          name: kritis-test-github
```
where is this resourceRef defined?

pivotal-nader-ziada replied:
I think we might not need the `resourceRef`, will reconsider this and provide more details

bobcatfish commented on examples/pipelines/guestbook.yaml:
```diff
@@ -8,76 +8,99 @@ spec:
+        name: stagingRegistry      
+      type: image
+      url: gcr.io/demo-staging
+      digest: latest
```
i'm having a hard time understanding if this resource represents an "image" or if it represents an "image registry". if it was an image, I would think it might look something like this:

```yaml
    - name: stagingImage
      type: image
      url: gcr.io/demo-staging
      imageName: kritis
```

- added `imageName`
- removed `digest` (won't be known until it's built?)
- changed the name from `stagingRegistry` to `stagingImage` since this resource is an image, not just a registry?

pivotal-nader-ziada replied:
makes sense, will make these changes, it should be referring to an image

bobcatfish commented on config/samples/pipeline_v1beta1_pipeline.yaml:
```diff
@@ -7,46 +7,60 @@ spec:
+          outputSourceBindings:
+            - name: stagingRegistry
+              sourceKey: stagingRegistry
+              passedConstraint: [build-push-task]
```
to fully understand the new design, I think the Tasks these are referring to need to be updated as well, for example in [integration-test-wizzbang-task](https://github.com/knative/build-pipeline/blob/master/examples/test_tasks.yaml) right now we've got:

```yaml
    inputs:
        sources:
            - name: workspace
        params:
            - name: testImage
              type: string
            - name: testArgs
              type: string
    outputs:
        results:
            - name: testResults
              format: junit
              path: integration/logs/tests.xml
```

But it looks like the `outputs` should be updated to indicate that the Task will be requiring an Output resource of type `image`?

pivotal-nader-ziada replied:
we updated some of them, must have missed this one, will update. You are correct in that the task needs an output resource of image if it is creating this image and passing it along.

tejal29 commented on README.md:
```diff
@@ -125,3 +123,16 @@ completes (or fails).
+* `Resources` is created directly in a pipeline configuration and bound 
+to `TaskRun` as an input and/or output source. 
+* The (optional) `passedConstraint` key on an `input source` defines a set of previous task names.
+* When the `passedConstraint` key is specified on an input source, only the version of 
```
Having a hard time wrapping my head around “passedConstraints” bound to “source”
Example: https://github.com/knative/build-pipeline/blob/cf8546cc88b6ee298f99b6f64b63d5db8acb78c2/examples/pipelines/guestbook.yaml#L79
Specifically in this case, we have 1 source key but 2 `passedConstraints`. 
Which source version do i pick?


I see a CI Pipeline defined like this:
Where, sources are input to a pipeline and then something comes out of the pipeline. It could be a production ready image or an app.
Pipeline consists of tasks and each task executes some steps on the source and then produce resources like (images, jars, test results) etc.
The pipeline will be run, managed by a controller.  Let us say we call it PipelineController.

[In my head Pipeline looks like this](https://docs.google.com/document/d/1nIxVqZbOLIPVRGuVvCEw3zn30l5fbvPgbwGrsLJS-dg/edit#heading=h.vgrkr0rpzh9d)
Each PipelineRun will have a RunContext which will store the state of the pipeline.
The PipelineController will add the inputs and outputs produced along the way are added to RunContext along with their versions.
Now, when Task1 completes, the PipelineController should invoke Task2. It will refer the versions of the inputs and outputs in the context and pass the required artifacts.
This behaviour should be “implicit”. 
We do not want different versions of the same source to be running in the same pipeline run.
Adding a `passingConstraint` which is optional at the source level may actually lead to such scenarios. 
Hence, i feel moving `passedConstraint` to task level makes more sense.

tejal29 replied:
Thanks for clarifying @pivotal-nader-ziada 

> We are trying to make it explicit that the version of the resource stays the same across multiple tasks in a pipelineRun. By adding the passedConstraint key on the InputSource that means that the input of that task has to be the same version that passed the previous task(s) mentioned in constraint.
This allows for the option of another resource (for example a test only repo) that always wants to get latest in each run and doesn't need the same version

Ok, this makes things more clear now. I think,  we just have defaults flipped.  If you have a resource defined like this, i was more inclined towards overriding the resource version.  
```
resource:
    - name: guestbook
      type: git
      url: github.com/kubernetes/examples
      serviceAccount: githubServiceAccount
      revision: HEAD
```
e.g. The first task in the pipeline pulling in guestbook will use revision `HEAD` and lock its sha. The next tasks in pipeline which depend on resource `guestbook` will override the revision with the locked sha built by 1st task in the pipeline.
I like your approach where you explicitly specify the `passedConstraint`. 
However, we are now relied on pipeline owners to make sure they correctly configure it. We can work around it by adding some "intelligence" in validators and spinning out warning if we see patterns where pipeline-creators are pulling in the same resource in two tasks but not using `passedConstraint`.

pivotal-nader-ziada commented:
@bobcatfish @tejal29 we addressed the comments

bobcatfish commented on config/samples/pipeline_v1beta1_pipelinerun.yaml:
```diff
@@ -14,16 +14,8 @@ status:
-      nextTasks: []
-      prevTasks:
-        - taskRef:
-          name: test-wizzbang-task-run-sd8f8dfasdfasdfas
```
i'd still like to see some representation in the PipelineRun of the order the Tasks were executed in, any thoughts on how to do that in the new design?

pivotal-nader-ziada replied:
The passedConstraint controls that. If there are no constraints that means all tasks can run in any order or in parallel. Do you mean you want to see the actual order that happened in runtime?

bobcatfish replied:
> Do you mean you want to see the actual order that happened in runtime?

Yes, exactly - especially since the failure of a Task would halt the execution of the rest of the Tasks, so it would be useful to see what actually executed and in what order

bobcatfish commented on README.md:
```diff
@@ -125,3 +123,16 @@ completes (or fails).
+* When the `passedConstraint` key is specified on an input source, only the version of 
+the resource that passed through the defined list of tasks is used.
+* The `passedConstraint` allows for `Tasks` to fan in and fan out, and ordering can be expressed explicitly 
+using this key since a task needing a resource from a another task would have to run after.
```
Any interest in updating the README Pipeline overview image to the one you presented today @shashwathi @pivotal-nader-ziada ? And adding the resources v. pipelineparams image?

(I'm happy to add these if you'd prefer)

pivotal-nader-ziada replied:
The same diagrams are in the PR branch. Should be updated. Feel free to add them in any other location you prefer

bobcatfish commented on examples/pipelines/guestbook.yaml:
```diff
@@ -8,76 +8,92 @@ spec:
           params:
               - name: pathToDockerfile
-                value: guestbook-go/Dockerfile
+                type: guestbook-go/Dockerfile
```
something must have gone wrong here :thinking: this should definitely be value

pivotal-nader-ziada replied:
Yeah didn’t look right as type but should be easy enough to fix either in this PR or as a follow up

pivotal-nader-ziada commented:
@bobcatfish if you think this PR has enough things right, can we merge it and add issues for the small things remaining. The comments are getting hard to keep track of.

bobcatfish commented:
THIS IS LOOKING GREAT!

I looked through and I have a few outstanding things, but I suggest we go ahead and merge this (whenever you are ready) and we can make some incremental updates afterward!

My biggest outstanding issues:

* I'm still feeling strongly that `resources` should be outside of `Pipeline` somewhere - from the way the new CRDs are defined, I would actually expect to see an example `resources.yaml` in [the examples dir](https://github.com/knative/build-pipeline/tree/ee2b9536cc7e0530a50019584ba59845e7d1ecd9/examples), and see a reference to that instance of Resources in [the example PipelineRun](https://github.com/knative/build-pipeline/blob/ee2b9536cc7e0530a50019584ba59845e7d1ecd9/examples/invocations/kritis-pipeline-run.yaml)
* I think that the `Tasks` themselves should change slightly since we unified artifacts and sources, e.g. [this example task](https://github.com/knative/build-pipeline/blob/ee2b9536cc7e0530a50019584ba59845e7d1ecd9/examples/build_task.yaml) has:

  ```yaml
  spec:
      inputs:
          sources:
              - name: workspace
          params:
              - name: pathToDockerFile
                type: string
      outputs:
          artifacts:
              - name: builtImage
                type: image
                storeKey: registry
  ```
  I feel like the artifacts section should change, since artifacts aren't really a thing?
* Not all examples have test coverage, so I think some are updated and some aren't, but that's really a problem of the test setup which I'll be fixing in #20 so I don't mind if you want to leave it as is, and I'll get to #20 asap!

/approve
/lgtm
/hold

^^ feel free to decide to remove the `/hold` whenever you're ready @pivotal-nader-ziada

bobcatfish commented:
(I'm gonna look into that CLA issue right now as well)

bobcatfish commented:
Okay I looked into the CLA issue and I noticed that in general there have been some problems in the past when the author + committer are different (https://github.com/knative/serving/issues/1439). Here are all my theories about what might be going wrong, I think (1) is more likely:

1. I noticed that in this PR https://github.com/knative/serving/pull/1966/ you both did something similar, but the committer and the author were the same, it was just the `Signed-of-by` that differed:
  ![image](https://user-images.githubusercontent.com/432502/45789118-f6373500-bc32-11e8-83f3-88ebc69bf110.png) So maybe making the author + committer the same would work? It also might be important that they match the person who opened the PR
2. I noticed that @shashwathi you're using a different email than you have in past commits, in the build repo your commits are under `shashwathireddy@gmail.com`, and in this repo they are under `sreddy@pivotal.io`, is it possible that the pivotal email somehow hasn't signed the CLA? (note I've heard that if the domain part of the emails doesn't match, the CLA bot gets upset so I'm not sure this will help)

Sorry about this, none of these answers are great, and it seems like the CLA bot does not work well with your pair programming workflow :( Let me know if you're able to work around it using any of the guesses above, and I'll continue to follow up internally to see if this can be improved.

p.s. [It seems like github has introduced its own mechanism for co-authoring commits which maybe wouldn't have these issues?](https://help.github.com/articles/creating-a-commit-with-multiple-authors/)

shashwathi commented:
/check-cla

pivotal-nader-ziada commented:
/hold cancel

pivotal-nader-ziada commented:
@bobcatfish 
 * we fixed the cla by force pushing the branch with just one author 
 * split out the resource
 * changed the task input and output

everything looks good from our side, can you take a look if we missed anything and approve again since we had a new commit

we can add follow up issues for more changes

Thanks

bobcatfish commented on config/samples/pipeline_v1beta1_resources.yaml:
```diff
@@ -0,0 +1,19 @@
+    type: image
+    params:
+    - name: url
+      value: gcr.io/wizzbang-staging
```
cool, i like that this is separate!

bobcatfish commented on config/samples/pipeline_v1beta1_pipeline.yaml:
```diff
@@ -5,48 +5,56 @@ metadata:
+        name: wizzbang
+    - name: wizzbangStagingImage
+      resourceRef:
+        name: wizzbangStagingImage   
```
i think that this reference belongs in the `PipelineRun`, since that would be the place where you'd bring everything together (otherwise everytime you use this Pipeline, you have to use the same resources). but we can iterate on this!

bobcatfish commented:
Looking good! I'm going to lgtm this so we can go ahead and merge it :D 

/lgtm

Thanks for fixing up the author issues @pivotal-nader-ziada , sorry about that :( now it looks like @shashwathi isn't officially an author of any commits :'(

How much does this interfere with your workflow? I want to follow up about this internally. Is there any chance you'd consider making use of something like [github's co-author feature](https://help.github.com/articles/creating-a-commit-with-multiple-authors/)?

bobcatfish commented:
Oh I'm also going to hit "update branch"

bobcatfish commented:
/lgtm

NOTHING REALLY CHANGED @knative-prow-robot CMON

shashwathi commented:
> How much does this interfere with your workflow? I want to follow up about this internally. Is there any chance you'd consider making use of something like github's co-author feature?

Yeah we can try github co-author next time.

willnorris commented:
I know we added support for GitHub co-author to our CLA checker, but I actually think it'll run into the same problem... it'll still complain about needing consent of the secondary author.  We can handle this case of different author and committer (or even co-author) when the two individuals are covered by the same CLA.  In fact, I'm pretty sure it was for Pivotal specifically that we added this feature.

I'll follow up privately to get some more information and try to debug what happened here.  In this case in particular, there might be an easy fix.  However, this is a use case we need to handle in the general case as well, but that's going to take a bit more time.

jglick commented on examples/pipelines/kritis-resources.yaml:
```diff
@@ -0,0 +1,26 @@
+    - name: revision
+      value: master
+    - name: url
+      value: https://github.com/grafeas/kritis-test  
```
This repository does not seem to exist. Was it subsequently deleted? Or not publicly readable?

bobcatfish replied:
funny you should mention this, i just noticed it myself! in #217 it's being updated to use a repo that exists - I think this was supposed to be an example of a case where someone has decided to put their tests in a separate repo (not something I'd recommend!)
